#!/bin/env snakemake -s

# Rules to generates the build bowtie2 index and split fastq for alignement.

GENOME = "/home/sardine/Documents/genomes/Sc288_2m/SC288_with_micron.fa"
R1 = "/home/sardine/Documents/reads/AC1/AC1.end1_sub.fastq"
R2 = "/home/sardine/Documents/reads/AC1/AC1.end2_sub.fastq"

OUTPUT = "/home/sardine/Bureau/"
BANK_NAME = "hic_test"
THREADS = 16
RATE = 0.2
MODE = "random"
# START = "fastq"
# EXIT = "bam"
ROUNDS = 10
MAGNITUDE = 0.8
SENSITIVITY = "very-fast"
ENZYMES_0 = "DpnII"
ENZYMES_1  = "HinfI"

# Set parameters.
shell.prefix("set -euo pipefail;")

# Specify at least one output of each rule (or one of the last output)  to ensure rule execution.
rule all:
    input: 
        "/home/sardine/Bureau/hic_test/index/",
        "/home/sardine/Bureau/hic_test/group0.1.bam",


rule hicberg_step_0:
    input: 
            genome = GENOME,
            r1 = R1,
            r2 = R2,
    output: 
            index_folder = directory("/home/sardine/Bureau/hic_test/index/"),
            for_sam = "/home/sardine/Bureau/hic_test/1.sam",
            rev_sam = "/home/sardine/Bureau/hic_test/2.sam",

    shell: 
        """
        hicberg pipeline -g {input.genome} --fq-for {input.r1} --fq-rev {input.r2} -o {OUTPUT} -r {RATE}  -t {THREADS} -m {MODE}  -e {ENZYMES_0} \
        -e {ENZYMES_1} -s {SENSITIVITY} -n {BANK_NAME} -R {ROUNDS} -M {MAGNITUDE} -f --start-stage fastq  --exit-stage bam
        """

rule hicberg_step_1:
    input:
        genome = GENOME,
        r1 = R1,
        r2 = R2,
        for_sam = "/home/sardine/Bureau/hic_test/1.sam",
        rev_sam = "/home/sardine/Bureau/hic_test/2.sam",

    params:
        main_directory = directory("/home/sardine/Bureau/hic_test/"),

    output: 
        unmapped_for_bam = "/home/sardine/Bureau/hic_test/group0.1.bam",
        unmapped_rev_bam = "/home/sardine/Bureau/hic_test/group0.2.bam",
        mapped_for_bam = "/home/sardine/Bureau/hic_test/group1.1.bam",
        mapped_rev_bam = "/home/sardine/Bureau/hic_test/group1.2.bam",
        multi_mapped_for_bam = "/home/sardine/Bureau/hic_test/group2.1.bam",
        multi_mapped_rev_bam = "/home/sardine/Bureau/hic_test/group2.2.bam"

    shell: 
        """
        hicberg pipeline -g {input.genome} --fq-for {input.r1} --fq-rev {input.r2} -o {OUTPUT} -r {RATE} \
        -t {THREADS} -m {MODE}  -e {ENZYMES_0} -e {ENZYMES_1} -s {SENSITIVITY} -n {BANK_NAME} -R {ROUNDS} \
        -M {MAGNITUDE} -f --start-stage bam  --exit-stage stats
        """
